src/System.Text.Json/tests/JsonDocumentTests.cs

```diff
@@ -3034,6 +3045,124 @@ public static void EnsureResizeSucceeds()
+            Assert.True(jElement.ValueEquals("".AsSpan()));
+            Assert.True(jElement.ValueEquals(default(ReadOnlySpan<char>)));
+            Assert.True(jElement.ValueEquals(default(ReadOnlySpan<byte>)));
+        }
+
+        [Fact]
+        public static void ValueEquals_TestTextEqualsLargeMatch_True()
+        {
+            var jsonChars = new char[320];  // Some value larger than 256 (stack threshold)
+            jsonChars.AsSpan().Fill('a');
+
+            Span<char> lookupChars = new char[jsonChars.Length];
+            jsonChars.CopyTo(lookupChars);
+
+            string jsonString = "\"" + new string(jsonChars) + "\"";
+            JsonElement jElement = (JsonElement)JsonDocument.Parse(jsonString).RootElement.Clone();
+
+            Assert.True(jElement.ValueEquals(new string(jsonChars)));
+            Assert.True(jElement.ValueEquals(jsonChars));
+            Assert.True(jElement.ValueEquals(lookupChars));
+        }
+
+        [Theory]
+        [InlineData("\"conne\\u0063tionId\"", "connectionId")]
+        [InlineData("\"connectionId\"", "connectionId")]
+        [InlineData("\"123\"", "123")]
+        [InlineData("\"My name is \\\"Ahson\\\"\"", "My name is \"Ahson\"")]
+        public static void ValueEquals_JsonTokenStringType_True(string jsonString, string expected)
+        {
+            JsonElement jElement = (JsonElement)JsonDocument.Parse(jsonString).RootElement.Clone();
+            Assert.True(jElement.ValueEquals(expected));
+            Assert.True(jElement.ValueEquals(expected.AsSpan()));
+            byte[] expectedGetBytes = Encoding.UTF8.GetBytes(expected);
+            Assert.True(jElement.ValueEquals(expectedGetBytes));
+        }
+
+        [Theory]
+        [InlineData("\"conne\\u0063tionId\"", "c")]
+        public static void ValueEquals_DestinationTooSmallComparesEscaping_False(string jsonString, string other)
+        {
+            JsonElement jElement = (JsonElement)JsonDocument.Parse(jsonString).RootElement.Clone();
+            Assert.False(jElement.ValueEquals(other));
+            Assert.False(jElement.ValueEquals(other.AsSpan()));
+            byte[] otherGetBytes = Encoding.UTF8.GetBytes(other);
+            Assert.False(jElement.ValueEquals(otherGetBytes));
+        }
+
+        [Theory]
+        [InlineData("\"hello\"", new char[1] { (char)0xDC01 })]    // low surrogate - invalid
+        [InlineData("\"hello\"", new char[1] { (char)0xD801 })]    // high surrogate - missing pair
+        public static void InvalidUTF16Search(string jsonString, char[] lookup)
+        {
+            JsonElement jElement = (JsonElement)JsonDocument.Parse(jsonString).RootElement.Clone();
+            Assert.False(jElement.ValueEquals(lookup));
+        }
+
+        [Theory]
+        [InlineData("\"connectionId\"", "\"conne\\u0063tionId\"")]
+        [InlineData("\"conne\\u0063tionId\"", "bonnectionId")] // intentionally changing the expected starting character
```

Comment:
Based on the implementation, there should also be a test for a mismatch after the first escaped character